CRM-21718: Don't choose 'print_contacts' as default task if it is not there

johanv opened the pull request:
Overview
----------------------------------------
If you remove 'print contacts' from the search tasks by implementing hook_civicrm_searchTasks, every search form stops working.

Before
----------------------------------------
If you remove 'print contacts' from the search tasks by implementing hook_civicrm_searchTasks, every search form stops working.
![screenshot from 2018-01-27 20-33-30](https://user-images.githubusercontent.com/586145/35475583-70b49bc4-03a1-11e8-94f5-3d75c01528a0.png)

After
----------------------------------------
Now it works.
![screenshot from 2018-01-27 20-36-23](https://user-images.githubusercontent.com/586145/35475603-c631416a-03a1-11e8-84b8-b4256d46bc60.png)

Technical Details
----------------------------------------
It seems that in some part of the code for some reason the 'print contacts' is selected as default action, which causes troubles if it is not there.

---

 * [CRM-21718: If you remove 'Print selected rows' using hook_civicrm_searchTasks, search forms stop working](https://issues.civicrm.org/jira/browse/CRM-21718)

mattwire commented:
@johanv I think this will affect all search tasks, not just contacts.  Could you wait until #11240 is merged and then refactor this?

johanv commented:
Ok, just ping me when I should look at this again.

eileenmcnaughton commented:
@johanv Can you test the latest rc - I believe this is now fixed due to @mattwire's work. Closing this

johanv commented:
Thank you @eileenmcnaughton and @mattwire for your work. Sorry that it took me so long to look at this again, but since I changed jobs I don't have a lot of time for CiviCRM. 
But tonight I finally found some time to look at this again. Unfortunately #11240 does not solve my issue.

So, I created an extension, and I implemented `hook_civicrm_searchTasks` like this:
```
function test_civicrm_searchTasks($objectType, &$tasks) {
  if ($objectType != 'contact') {
    return;
  }

  foreach ($tasks as $key => $value) {
    if ($value['title'] == 'Print selected rows') {
      unset($tasks[$key]);
    }
  }

  return;
}
```
In CiviCRM 4.7.29 with the patch in this pull request, this would remove the item 'Print selected rows' from the search tasks.
With the current CiviCRM master branch, this does nothing.
I used my debugger, and it turned out that in the current master branch, `$objectType` is `null`. So to test, I removed the first `return` statement, and in that case, I got this error message:
```
 Fatal error: require_once(): Failed opening required '.php' (include_path='/opt/ext/be.johanv.test/:/var/www/html/sites/all/modules/civicrm/tools/extensions/cividiscount/:/var/www/html/sites/all/modules/civicrm/tools/extensions/civisualize/:/var/www/html/sites/all/modules/civicrm/tools/extensions/org.civicoop.civirules/:/var/www/html/sites/all/modules/civicrm/tools/extensions/civivolunteer/:/var/www/html/sites/all/modules/civicrm/tools/extensions/org.civicrm.angularprofiles/:.:/var/www/html/sites/all/modules/civicrm/:/var/www/html/sites/all/modules/civicrm//packages:.:/var/www/html/sites/all/modules/civicrm:/var/www/html/sites/all/modules/civicrm/packages:/var/www/html/sites/all/modules/civicrm/vendor/tecnickcom:/var/www/html/sites/all/modules/civicrm/vendor/phpseclib/phpseclib/phpseclib:/var/www/html/sites/all/modules/civicrm/vendor/pear/pear_exception:/var/www/html/sites/all/modules/civicrm/vendor/pear/auth_sasl:/var/www/html/sites/all/modules/civicrm/vendor/pear/net_socket:/var/www/html/sites/all/modules/civicrm/vendor in /var/www/html/sites/all/modules/civicrm/CRM/Core/Controller.php on line 451
```
![screenshot from 2018-03-19 22-09-00](https://user-images.githubusercontent.com/586145/37622539-34d2b070-2bc2-11e8-939f-3a186d22819a.png)

Unfortunately, I don't have time to look further into this tonight. Maybe another time.

eileenmcnaughton commented:
@johanv @mattwire please test #11861
